sdk/dotnet/Pulumi/Deployment/TaskMonitoringHelper.cs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Pulumi
+{
+    /// Monitors dynamically added tasks for completion. Enters IDLE
+    /// state when all monitored tasks finish. Allows awaiting next
+    /// IDLE state or the first exception, whichever comes first.
+    /// Thread-safe.
+    internal class TaskMonitoringHelper
+    {
+        private readonly object _lockObject = new Object();
+        private int _activeTasks = 0;
+        private TaskCompletionSource<Exception?>? _next = null;
```

Comment:
Nit: There's no need to specify the default values (smaller IL):

```suggestion
        private int _activeTasks;
        private TaskCompletionSource<Exception?>? _next;
```